### PR TITLE
Simplify maybe_resize_out

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88117
* __->__ #88116
* #88115

The previous behaviour would call `resize_` on 0-sized elements even
when their size was correct. This would make some test fail, as resize_
may be an in-place operation and it's not supported by some subsystems

cc @ezyang @mruberry @ngimel @Lezcano @peterbell10 @fdrocha

### CODE REVIEW

File: torch/_prims_common/wrappers.py [hunk 1]
```diff
@@ -130,6 +130,9 @@ def _fn(*args, **kwargs):
 
 # TODO: handle tuples of tensors
 def _maybe_resize_out(out: TensorLikeType, shape):
+    if out.shape == shape:
```
Comment (mruberry) [inline torch/_prims_common/wrappers.py:133]: This short-circuit seems fine. Let's add a comment explaining it.

File: torch/_prims_common/wrappers.py [hunk 2]
```diff
@@ -130,6 +130,9 @@ def _fn(*args, **kwargs):
...
+        return out
+
```
Comment (mruberry) [inline torch/_prims_common/wrappers.py:135]: Isn't it weird that the check below, on line 139, is for whether the number of elements in the `out` tensor and the shape match? The error message says they must have the same shape, not the same number of elements. Also, while we're here, shouldn't we update it to use something like `utils.warn`, analogous to `utils.check` but for warnings?

Comment (lezcano) [reply torch/_prims_common/wrappers.py:135]: I think that, given that we don't use warns all that much, it's fine to leave it as it is for now.

As for the logic, I simplified it a bit (not pushed yet).

File: torch/_prims_common/wrappers.py [hunk 3]
```diff
@@ -130,24 +130,21 @@ def _fn(*args, **kwargs):
...
-    if out.numel() != reduce(operator.mul, shape, 1):
-        msg = (
-            "An output with one or more elements was resized since it had shape {0} "
-            "which does not match the required output shape {1}. "
-            "This behavior is deprecated, and in a future PyTorch release outputs will not "
-            "be resized unless they have zero elements. "
-            "You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0).".format(
-                str(out.shape), str(shape)
+    # If the shapes are correct there's nothing to do
+    if out.shape == shape:
```
Comment (mruberry) [inline torch/_prims_common/wrappers.py:134]: Nit: use `utils.same_shape`

https://github.com/pytorch/pytorch/blob/dc40d3f93f849e467b2b56595a01f28e84ac7fa2/torch/_prims_common/__init__.py#LL89-L97

**Review by mruberry — APPROVED** [review]
Cool -- see one inline suggestion

**Pull request closed without merging**